service/service.go

```diff
@@ -57,7 +58,9 @@ func (svc *Service) IsTombstone() bool {
 }
 
 func (svc *Service) Invalidates(otherSvc *Service) bool {
-	return otherSvc != nil && svc.Updated.After(otherSvc.Updated)
+	return otherSvc != nil &&
+		svc.Updated.After(otherSvc.Updated) &&
+		!(otherSvc.Status == DRAINING && svc.Status == ALIVE)
```

Comment:
We really don't want to have an exception here for `DRAINING`. This should just be a normal status like any other. If it's newer, it should take precedence. Things in `DRAINING` state need to not announce `ALIVE` status, so it should be fine without this. Any exceptions here can lead to stuck state, so it's dangerous unless we really have to do that.

Reply:
Looking at this more closely, I think you are right that this has to be done this way. The other only way is to mess with Discovery which I don't think is a good solution.

Reply:
@relistan Actually, I thought I was being smart by doing it that way, but it didn't really work, since it still needs to call `state.ServiceChanged` and update the timestamps. Otherwise [`TombstoneOthersServices`](https://github.com/Nitro/sidecar/blob/8912f899907a44f2c1ca969d00aeff7422b8ad61/catalog/services_state.go#L593) thinks it expired.

I ended up implementing this logic in AddServiceEntry and it looks like this does work as expected now. See c89de8a.